ui/app/src/state/reducers/previewNavigation.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2007-2021 Crafter Software Corporation. All Rights Reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { createReducer } from '@reduxjs/toolkit';
+import GlobalState from '../../models/GlobalState';
+import { envInitialState } from './env';
+import {
+  changeCurrentUrl,
+  checkInGuest,
+  goToLastPage,
+  goToNextPage,
+  GUEST_CHECK_IN,
+  GUEST_CHECK_OUT
+} from '../actions/preview';
+import { changeSite } from './sites';
+
+const guestBase = envInitialState.guestBase;
+const previewLanding = envInitialState.previewLandingBase;
+
+function cleanseUrl(url: string) {
+  const clean = url || '/';
+  if (!clean.startsWith('/')) {
+    return `/${clean}`;
+  }
+  return clean;
+}
+
+const reducer = createReducer<GlobalState['previewNavigation']>(
+  {
+    // What's shown to the user across the board (url, address bar, etc)
+    currentUrlPath: '',
+    // The src of the iframe
+    currentFullUrl: previewLanding,
+    historyBackStack: [],
+    historyForwardStack: [],
+    historyNavigationType: null
+  },
+  {
+    [checkInGuest.type]: (state, { payload }) => {
```

Comment:
Please add these type of comments to each action
```suggestion
    // A page is being visited or reloaded
    // - Push the url into the back stack (if not already pushed)
    // - Clear the forward stack (if not already pushed)
    // - Update the current url
    [checkInGuest.type]: (state, { payload }) => {
```

Reply:
Still missing the comments for the rest of the actions